### PR TITLE
recall only cards not already in holder

### PR DESCRIPTION
**Pull request by robartsd**
Recall of decks currently causes all cards to move to the holder even if they are already there. This change should noticeably improve performance when recalling large decks that are mostly already in the holder.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3537/pr-test (or any other room on that server)

**Comment by robartsd** [comment]
Seems to be working now. Some TestCafe tests are failing (presumably because recalled cards end up in a different order - or possibly just with different z values - will check later).

**Comment by robartsd** [comment]
TestCafe tests involving recall had some state changes. Many had most or all of the deck in different state because the order of the cards after recall and before shuffle was different. Reward only recalled markers, but many ended up with different z values. SORT was interesting because all changes were reducing card z-values by 108. I did get weird results in the diff for Undercover (lots of reference lines and a few blocks of routine code that was listed as changes, but lines were identical):

```
 Running tests in:
 - Chrome 91.0.4472.77 / Linux 0.0

 virtualtabletop.io
 {
   Reset: {
     clickRoutine: [
       ...
       ...
       "var redBegins = randInt 0 1"
       "var blueBegins = 1 - ${redBegins}"
       ...
       ...
       ...
       "var reds = 8 + ${redBegins}"
       "var blues = 8 + ${blueBegins}"
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
       ...
-      {
-        func: "SET"
-        property: "owner"
-        value: null
-      }
+      {
+        func: "SET"
+        property: "owner"
+        value: null
+      }
       ...
       ...
       ...
       ...
       ...
       ...
-      {
-        func: "SET"
-        property: "owner"
-        value: null
-      }
+      {
+        func: "SET"
+        property: "owner"
+        value: null
+      }
       ...
       ...
     ]
   }
   Spy Master Button: {
     clickRoutine: [
       ...
       ...
       "var owners = push ${playerName}"
       ...
       "var spyMasterCount = ${owners} length"
       "var spyMastersNeeded = ${spyMasterCount} != 2"
       ...
       ...
       ...
       ...
       "var console = from ${owners}"
       "var console = 'Chefs\\u003a' unshift"
       "var console = push ''"
       "var console = push 'Startteam\\u003a'"
       ...
       ...
       "var skipRed = 10 != ${COUNT}"
       "var skipBlue = 10 == ${COUNT}"
       ...
       ...
-      {
-        note: "This was added by the automatic file migration because the new expression syntax does not support skip."
-        func: "SET"
-        collection: "thisButton"
-        property: "internal_computeMigration_skip"
-        value: null
-      }
+      {
+        note: "This was added by the automatic file migration because the new expression syntax does not support skip."
+        func: "SET"
+        collection: "thisButton"
+        property: "internal_computeMigration_skip"
+        value: null
+      }
       ...
       ...
-      {
-        note: "This was added by the automatic file migration because the new expression syntax does not support skip."
-        func: "SET"
-        collection: "thisButton"
-        property: "internal_computeMigration_skip"
-        value: null
-      }
+      {
+        note: "This was added by the automatic file migration because the new expression syntax does not support skip."
+        func: "SET"
+        collection: "thisButton"
+        property: "internal_computeMigration_skip"
+        value: null
+      }
       "var spyMastersList = ${console} join '\\u000a'"
       ...
       ...
-      {
-        func: "SET"
-        property: "owner"
-        value: null
-        skip: "${spyMastersNeeded}"
-      }
+      {
+        func: "SET"
+        property: "owner"
+        value: null
+        skip: "${spyMastersNeeded}"
+      }
       ...
-      {
-        func: "SET"
-        property: "owner"
-        value: null
-        skip: "${spyMastersNeeded}"
-      }
+      {
+        func: "SET"
+        property: "owner"
+        value: null
+        skip: "${spyMastersNeeded}"
+      }
       ...
       ...
     ]
   }
   cxzl: {
-    z: 979
+    z: 8450
   }
   yi8y: {
-    z: 7888
+    z: 9750
   }
   bagc: {
-    z: 1848
+    z: 7888
   }
   gsvd: {
-    z: 5824
+    z: 7572
   }
   vsq5: {
-    z: 8659
+    z: 979
   }
   kxaq: {
-    z: 8450
+    z: 5824
   }
   05ug: {
-    z: 1275
+    z: 2708
   }
   habk: {
-    z: 9668
+    z: 1275
   }
   dh8t: {
-    z: 8784
+    z: 6703
   }
   0va5: {
-    z: 735
+    z: 9668
   }
   y9rl: {
-    z: 6703
+    z: 735
   }
   tlzs: {
-    z: 2708
+    z: 8784
   }
   qhk5: {
-    z: 4869
+    z: 4525
   }
   2ndn: {
-    z: 7572
+    z: 1848
   }
   5kpy: {
-    z: 9750
+    z: 8659
   }
   xjsv: {
-    z: 4525
+    z: 4869
   }
   llf4: {
-    z: 12569
+    z: 12572
-    parent: "9pg9"
+    parent: "q62e"
   }
   2elx: {
-    z: 40842
+    z: 40793
-    parent: "q62e"
+    parent: "9pg9"
   }
   0v0h: {
-    z: 40840
+    z: 40792
   }
   9rj5: {
-    z: 12571
+    z: 12570
   }
 }

 ✖ Public library: Undercover (variant 1)

   1) AssertionError: expected '3e9afb911afbcffd5447ac27f8d0b10f' to deeply equal '1c27b314acacb62fa3543f7aee49aacd'
```

**Review by ArnoldSmith86 — DISMISSED** [review]
Load `The Team` (3 players), click `Deal`. One player receives a card (`xo1u`) from a deck that should actually be recalled.

Also `c.parent` is not always the parent you're looking for, I think. I would recommend `widgets.get(c.get('parent'))` instead.

**Comment by bjalder26** [comment]
It seems to work fine if none of the cards are in a pile when they are recalled.

**Comment by robartsd** [comment]
> It seems to work fine if none of the cards are in a pile when they are recalled.

Piles always seem to be the problem. I'll probably try to look at the trace that ArnoldSmith86 shared from the game this morning where it was discovered to see if I can make some sense of it.

**Comment by ArnoldSmith86** [comment]
Since it's easily reproducible the browser debugger is probably more helpful.

**Comment by bjalder26** [comment]
> > It seems to work fine if none of the cards are in a pile when they are recalled.
> 
> Piles always seem to be the problem. I'll probably try to look at the trace that ArnoldSmith86 shared from the game this morning where it was discovered to see if I can make some sense of it.

Just hit either of the two recall buttons when the cards belonging to the holder by them have cards stacked. http://212.47.248.129:3537/Interitus

**Comment by ArnoldSmith86** [comment]
In case it saves some time: the holder and deck in Team are off screen.

**Comment by bjalder26** [comment]
I might consider pushing PR #559 on this to see what problems remain after those fixes.

**Comment by ArnoldSmith86** [comment]
Aren't they both in the staging PR?

**Comment by bjalder26** [comment]
> Aren't they both in the staging PR?

Yes, and I think that is why in the staging PR it doesn't freeze every time you try to recall a card that is part of a stack.  So sure, we could look at the staging PR to 'fix' the 'remaining' issues with this PR as long as we are confident that the issues there are just from this PR.

File: client/js/widgets/widget.js
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:846]: This is causing the staging PR error. Should be `widgets.get`, right?

**Comment by robartsd** [comment]
Looks like the typo that ArnoldSmith86 found this morning was the only problem.

**Comment by bjalder26** [comment]
This does not appear to have the problem it had before.

**Comment by bjalder26** [comment]
There are still pile-related issues.  It probably isn't a reason to not approve this PR, but if one card is left in the holder, and the rest are recalled, then the one in the holder will 'leave' (lose the holder as its parent) as a pile is formed.  All the rest of the cards will enter, then leave as they get put on the pile.
image
To me **it looks like it is doing what it is supposed to do**, but as piles form the routines are triggered.

**Comment by ArnoldSmith86** [comment]
But the cards are then all in one pile in the holder? The leave and enter problems are unrelated IMO.

**Comment by robartsd** [comment]
> To me **it looks like it is doing what it is supposed to do**, but as piles form the routines are triggered.

Yes, that's a pile problem out of scope of this PR.

**Comment by robartsd** [comment]
Working on the PCIO recall PR, I realized that perhaps the check if the card is already in the holder would be better moved to the widget's moveToHolder method.

**Pull request closed without merging**
